### PR TITLE
Fix #347: Faster text rendering in Agg

### LINKED ISSUE #347: Faster Text drawing needed
[Original report at SourceForge, opened Wed Mar 21 08:03:30 2007](https://sourceforge.net/tracker/?func=detail&aid=1685117&group_id=80706&atid=560723)

Drawing of plots with multiple matplotlib.Text instances is so slow as to be unusable for some applications, such as realtime multiple sequence alignment rendering/viewing (see attachment for a pylab example, resize plot window to see effect).

Some layman's thoughts:
* Would it be possible to speed things up by implementing some kind of Text Collection?

* Is antialiasing involved? An option to turn it off would be nice in that case.

#### On Thu Sep  6 16:07:54 2007, mdboom wrote:
Logged In: YES 
user_id=119312
Originator: NO

It appears that the text is re-rendered to a newly allocated buffer each time the text is drawn.  Perhaps some caching of text elements would help.

(Mostly just making a note of it for someone else to work this through, or until I have the time.)

### SourceForge History

* On Wed Mar 21 08:03:30 2007, by yohell: File Added: 221568: test.py

### PR DESCRIPTION
**Pull request by mdboom**
When the text isn't rotated, we can skip a lot of machinery by just doing a direct blending of text pixels onto the output buffer.

With the `simple_plot_fps` demo, I get about 6fps after, 5fps before this change.

Cc: @Tillsten: would you be willing to kick the tires on this one?

### CODE REVIEW

**Comment by jkseppan** [comment]
Note to reviewers: because of the indentation change, the github diff is big, but you can see the effective diff with

    git remote add mdboom https://github.com/mdboom/matplotlib.git
    git fetch mdboom
    git show -w mdboom/fast-text

File: src/_backend_agg.h [hunk 1]
```diff
@@ -734,41 +734,68 @@ inline void RendererAgg::draw_text_image(GCAgg &gc, ImageArray &image, int x, in
...
-    span_gen_type output_span_generator(&image_span_generator, gc.color);
-    renderer_type ri(rendererBase, sa, output_span_generator);
-
-    theRasterizer.add_path(rect2);
-    agg::render_scanlines(theRasterizer, slineP8, ri);
+        for (int yi = text.y1; yi < text.y2; ++yi) {
+            for (int xi = text.x1; xi < text.x2; ++xi) {
+                typename ImageArray::value_type pixel = image(
+                        yi - (y - image.dim(0)), xi - x);
+                pixFmt.blend_pixel(xi, yi, gc.color, pixel);
```
Comment (jkseppan) [inline src/_backend_agg.h]: I'm not an agg expert, but it seems that blend_color_hspan could be even faster.

http://sourceforge.net/p/vector-agg/mailman/message/5116797/

**Comment by QuLogic** [comment]
> Note to reviewers: because of the indentation change, the github diff is big

You can also add `?w=1` to the end of the URL to have the view ignore whitespace changes.

File: src/_backend_agg.h (hunk 1 above)
Comment (mdboom) [inline src/_backend_agg.h]: I'll play with that.  `blend_color_hspan` doesn't take as inputs exactly the same thing that we have, so there may be overhead in creating that first negating any wins.  Only benchmarking will tell.

**Comment by mdboom** [comment]
I've updated this to use `blend_color_solid` (which is the single color version of `blend_color_hspan`).  It isn't measurably faster, but it's fewer lines of code for us.

**Comment by mdboom** [comment]
I'm remilestoning to 2.0, since this will affect image comparison tests (for the better).

**Comment by tacaswell** [comment]
cherry picked to v2.0.x as a340827

**Pull request merged**